docs/tutorials/onnx/fine_tuning_gluon.md

```diff
@@ -0,0 +1,420 @@
+
+
+```python
+pre_trained = gluon.nn.SymbolBlock(outputs=new_sym, inputs=mx.sym.var('input_0'))
+net_params = pre_trained.collect_params()
+for param in new_params:
+    if param in net_params:
+        net_params[param]._load_init(new_params[param], ctx=ctx)
+```
+
+We create the new dense layer with the right new number of classes (101) and initialize the weights
+
+
+```python
+dense_layer = gluon.nn.Dense(NUM_CLASSES)
+dense_layer.collect_params().initialize(mx.init.Xavier(magnitude=2.24), ctx=ctx)
+```
+
+We add the SymbolBlock and the new dense layer to a HybridSequential network
+
+
+```python
+net = gluon.nn.HybridSequential()
+net.add(pre_trained)
+net.add(dense_layer)
+```
+
+### Loss
+Softmax cross entropy for multi-class classification
+
+
+```python
+softmax_cross_entropy = gluon.loss.SoftmaxCrossEntropyLoss()
+```
+
+### Trainer
+Initialize trainer with common training parameters
+
+
+```python
+LEARNING_RATE = 0.001
+WDECAY = 0.00001
+MOMENTUM = 0.9
+```
+
+The trainer will retrain and fine-tune the entire network. If we use `dense_layer` instead of `net` in the cell below, the gradient updates would only be applied to the new last dense layer. Essentially we would be using the pre-trained network as a featurizer.
+
+
+```python
+trainer = gluon.Trainer(net.collect_params(), 'sgd', 
+                        {'learning_rate': LEARNING_RATE, 
+                         'wd':WDECAY, 
+                         'momentum':MOMENTUM})
+```
+
+### Evaluation loop
+
+We measure the accuracy in a non-blocking way, using `nd.array` to take care of the parallelisation that MXNet and Gluon offers.
+
+
+```python
+ def evaluate_accuracy_gluon(data_iterator, net):
+    num_instance = nd.zeros(1, ctx=ctx)
+    sum_metric = nd.zeros(1,ctx=ctx, dtype=np.int32)
+    for i, (data, label) in enumerate(data_iterator):
+        data = data.astype(np.float32).as_in_context(ctx)
+        label = label.astype(np.int32).as_in_context(ctx)
+        output = net(data)
+        prediction = nd.argmax(output, axis=1).astype(np.int32)
+        num_instance += len(prediction)
+        sum_metric += (prediction==label).sum()
+    accuracy = (sum_metric.astype(np.float32)/num_instance.astype(np.float32))
+    return accuracy.asscalar()
+```
+
+
+```python
+%%time
+print("Untrained network Test Accuracy: {}".format(evaluate_accuracy_gluon(dataloader_test, net)))
+```
+
+`Untrained network Test Accuracy: 0.01923076994717121`<!--notebook-skip-line-->
+
+
+
+### Training loop
+
+
+```python
+val_accuracy = 0
+for epoch in range(20):
+    for i, (data, label) in enumerate(dataloader_train):
+        data = data.astype(np.float32).as_in_context(ctx)
+        label = label.as_in_context(ctx)
+        with autograd.record():
+            output = net(data)
+            loss = softmax_cross_entropy(output, label)
+        loss.backward()
+        trainer.step(data.shape[0])
+    nd.waitall() # wait at the end of the epoch
```

Comment:
For performance reason, it is better to put this line right after label = label.as_in_context(ctx), so if a user has GPU loading of the data and calculation happens simultaneously

Reply:
This line is done at the end of the epoch, not at the end of the batch (so the entire dataset is loaded in memory at the beginning of the epoch already)